android-build-tools/src/apk.rs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    fn apk(&self) -> PathBuf {
+        self.build_dir
+            .join(format!("{}.apk", self.manifest.package_label))
+    }
+
+    pub fn create_apk(&self) -> Result<UnalignedApk, NdkError> {
+        std::fs::create_dir_all(&self.build_dir)?;
+        self.manifest.write_to(&self.build_dir)?;
+
+        let mut aapt = self.build_tool("aapt")?;
+        aapt.arg("package")
+            .arg("-f")
+            .arg("-F")
+            .arg(self.unaligned_apk())
+            .arg("-M")
+            .arg("AndroidManifest.xml")
+            .arg("-I")
+            .arg(self.ndk.android_jar(self.manifest.target_sdk_version)?);
+
+        if let Some(res) = &self.res {
+            aapt.arg("-S").arg(res);
+        }
+
+        if let Some(assets) = &self.assets {
+            aapt.arg("-A").arg(assets);
+        }
+
+        aapt.status()?;
+
+        Ok(UnalignedApk(self))
+    }
+}
+
+pub struct UnalignedApk<'a>(&'a ApkConfig);
+
+impl<'a> UnalignedApk<'a> {
+    pub fn add_lib(&self, path: &Path, target: Target) -> Result<(), NdkError> {
+        if !path.exists() {
+            return Err(NdkError::PathNotFound(path.into()));
+        }
+        let abi = target.android_abi();
+        let file_name = path.file_name().unwrap();
+        let out = self.0.build_dir.join("lib").join(abi);
+        std::fs::create_dir_all(&out)?;
+        std::fs::copy(path, out.join(&file_name))?;
+
+        self.0
+            .build_tool("aapt")?
+            .arg("add")
+            .arg(self.0.unaligned_apk())
+            .arg(format!("lib/{}/{}", abi, file_name.to_str().unwrap()))
+            .status()?;
+        Ok(())
+    }
+
+    pub fn align(self) -> Result<UnsignedApk<'a>, NdkError> {
+        self.0
+            .build_tool("zipalign")?
+            .arg("-f")
+            .arg("-v")
+            .arg("4")
+            .arg(self.0.unaligned_apk())
+            .arg(self.0.apk())
+            .status()?;
+        Ok(UnsignedApk(self.0))
+    }
+}
+
+pub struct UnsignedApk<'a>(&'a ApkConfig);
+
+impl<'a> UnsignedApk<'a> {
+    pub fn sign(self, key: Key) -> Result<Apk, NdkError> {
+        self.0
+            .build_tool("apksigner")?
+            .arg("sign")
+            .arg("--ks")
+            .arg(&key.path)
+            .arg("--ks-pass")
+            .arg(format!("pass:{}", &key.password))
+            .arg(self.0.apk())
+            .status()?;
+        Ok(Apk::from_config(self.0))
+    }
+}
+
+pub struct Apk {
+    path: PathBuf,
+    package_name: String,
+    ndk: Ndk,
+}
+
+impl Apk {
+    pub fn from_config(config: &ApkConfig) -> Self {
+        let ndk = config.ndk.clone();
+        Self {
+            path: config.apk(),
+            package_name: config.manifest.package_name.clone(),
+            ndk,
+        }
+    }
+
+    pub fn install(&self) -> Result<(), NdkError> {
+        self.ndk
+            .platform_tool("adb")?
+            .arg("install")
```

Comment:
The attempt to reinstall already installed app on older Android will fail with error like the follow:
```
Failure [INSTALL_FAILED_ALREADY_EXISTS]
```
As a simple workaround I suggest to add '-r' flag after 'install' command.